chain/network/src/routing.rs

```diff
@@ -0,0 +1,375 @@
+        if let Some(adj) = self.adjacency.get(&peer0) {
+            if adj.contains(&peer1) {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    fn add_directed_edge(&mut self, peer0: PeerId, peer1: PeerId) {
+        self.adjacency.entry(peer0).or_insert_with(HashSet::new).insert(peer1);
+    }
+
+    fn remove_directed_edge(&mut self, peer0: &PeerId, peer1: &PeerId) {
+        self.adjacency.get_mut(&peer0).unwrap().remove(&peer1);
+    }
+
+    pub fn add_edge(&mut self, peer0: PeerId, peer1: PeerId) {
+        if !self.contains_edge(&peer0, &peer1) {
+            self.add_directed_edge(peer0.clone(), peer1.clone());
+            self.add_directed_edge(peer1, peer0);
+        }
+    }
+
+    pub fn remove_edge(&mut self, peer0: &PeerId, peer1: &PeerId) {
+        if self.contains_edge(&peer0, &peer1) {
+            self.remove_directed_edge(&peer0, &peer1);
+            self.remove_directed_edge(&peer1, &peer0);
+        }
+    }
+
+    // TODO(MarX): This is too slow right now. (See benchmarks)
+    /// Compute for every node `u` on the graph (other than `source`) which are the neighbors of
+    /// `sources` which belong to the shortest path from `source` to `u`. Nodes that are
+    /// not connected to `source` will not appear in the result.
+    pub fn calculate_distance(&self) -> HashMap<PeerId, HashSet<PeerId>> {
+        let mut queue = vec![];
+        let mut distance = HashMap::new();
+        // TODO(MarX): Represent routes more efficiently at least while calculating distances
+        let mut routes: HashMap<PeerId, HashSet<PeerId>> = HashMap::new();
+
+        distance.insert(&self.source, 0);
+
+        // Add active connections
+        if let Some(neighbors) = self.adjacency.get(&self.source) {
+            for neighbor in neighbors {
+                queue.push(neighbor);
+                distance.insert(neighbor, 1);
+                routes.insert(neighbor.clone(), vec![neighbor.clone()].drain(..).collect());
+            }
+        }
+
+        let mut head = 0;
+
+        while head < queue.len() {
+            let cur_peer = queue[head];
+            let cur_distance = *distance.get(cur_peer).unwrap();
+            head += 1;
+
+            if let Some(neighbors) = self.adjacency.get(&cur_peer) {
+                for neighbor in neighbors {
+                    if !distance.contains_key(&neighbor) {
+                        queue.push(neighbor);
+                        distance.insert(neighbor, cur_distance + 1);
+                        routes.insert(neighbor.clone(), HashSet::new());
+                    }
+
+                    // If this edge belong to a shortest path, all paths to
+                    // the closer nodes are also valid for the current node.
+                    if *distance.get(neighbor).unwrap() == cur_distance + 1 {
+                        let adding_routes = routes.get(cur_peer).unwrap().clone();
```

Comment:
Don't you need to add an extra hop at the end of each such route before inserting them below?

Reply:
Oh, it's not hops, it's the first jump. Disregard my comments here.